File: Difficulty/Medium/[162] Find Peak Element/solution.cpp
/*
 * @lc app=leetcode id=162 lang=cpp
 *
 * [162] Find Peak Element
 */

// @lc code=start
class Solution {
public:
    int findPeakElement(vector<int>& nums) {
        if (nums.size() == 1) return 0;
        int left = 0, right = nums.size() - 1;
        while (left < right) {
            int mid = (left + right) / 2;
            int left_value = (mid - 1) >= 0 ? nums[mid - 1]: INT_MIN;
            int right_value = (mid + 1) < nums.size() ? nums[mid + 1]: INT_MIN;
            if (nums[mid] > left_value && nums[mid] > right_value)
                return mid;
            if (nums[mid] > left_value)
                left = mid + 1;
            else
                right = mid - 1;            
        }
        if (nums[0] > nums[1])
            return 0;
        if (nums[nums.size() - 1] > nums[nums.size() - 2])
            return nums.size() - 1;
        return (left + right) / 2;
    }
};
// @lc code=end

File: Difficulty/Medium/[162] Find Peak Element/solution.py
class Solution:
    def findPeakElement(self, nums: List[int]) -> int:
        if len(nums) == 1: return 0
        left, right = 0, len(nums) - 1
        while left < right:
            mid = (left + right) // 2
            left_value = nums[mid - 1] if (mid - 1) >= 0 else -2**31
            right_value = nums[mid + 1] if (mid + 1) < len(nums) else 2**31 - 1
            if nums[mid] > left_value and nums[mid] > right_value:
                return mid
            if nums[mid] > left_value:
                left = mid + 1
            else:
                right = mid - 1
        if nums[0] > nums[1]:
            return 0
        if nums[len(nums) - 1] > nums[len(nums) - 2]:
            return len(nums) - 1
        return (left + right) // 2